# Lab-02: Representación de Problemas, Búsqueda y Fundamentos Matemáticos de IA

## 4.2 Representación de un problema mediante estados

### Parte A: Ficha del problema del repartidor

**Estado inicial:** Almacén.

**Estado objetivo:** Zona Norte.

**Espacio de estados:** Almacén, Centro, Terminal, Mercado y Zona Norte.

**Acciones:** Desplazarse desde un punto de la ciudad hacia otro punto conectado directamente en el mapa.

**Costo de cada acción:** 1 paso por cada desplazamiento entre dos puntos conectados.


### Parte B: Representación del mapa como grafo

In [13]:
mapa = {
    "Almacen": ["Centro", "Terminal"],
    "Centro": ["Almacen", "ZonaNorte", "Mercado"],
    "Terminal": ["Almacen", "Mercado"],
    "Mercado": ["Centro", "Terminal", "ZonaNorte"],
    "ZonaNorte": ["Centro", "Mercado"],
}

print(mapa)

{'Almacen': ['Centro', 'Terminal'], 'Centro': ['Almacen', 'ZonaNorte', 'Mercado'], 'Terminal': ['Almacen', 'Mercado'], 'Mercado': ['Centro', 'Terminal', 'ZonaNorte'], 'ZonaNorte': ['Centro', 'Mercado']}


## 4.3 Implementación de BFS

In [14]:
def bfs(grafo, inicio, objetivo):
    """
    Busca el camino más corto entre 'inicio' y 'objetivo'
    usando búsqueda en amplitud (BFS).
    """

    cola = [[inicio]]
    visitados = set()

    while cola:
        camino = cola.pop(0)
        nodo_actual = camino[-1]

        if nodo_actual == objetivo:
            return camino

        if nodo_actual not in visitados:
            visitados.add(nodo_actual)

            for vecino in grafo[nodo_actual]:
                nuevo_camino = camino + [vecino]
                cola.append(nuevo_camino)

    return None
    

In [15]:
resultado_bfs = bfs(
    mapa,
    "Almacen",
    "ZonaNorte"
)

print("Camino encontrado con BFS:", resultado_bfs)

Camino encontrado con BFS: ['Almacen', 'Centro', 'ZonaNorte']


## 4.3. Implementación de DFS

DFS (Depth-First Search) explora un camino en profundidad antes de retroceder y probar otras alternativas.

A diferencia de BFS, utiliza una pila LIFO, donde el último camino agregado es el primero en salir.

In [16]:
mapa = {
    "Almacen": ["Centro", "Terminal"],
    "Centro": ["Almacen", "ZonaNorte", "Mercado"],
    "Terminal": ["Almacen", "Mercado"],
    "Mercado": ["Centro", "Terminal", "ZonaNorte"],
    "ZonaNorte": ["Centro", "Mercado"],
}

print(mapa)

{'Almacen': ['Centro', 'Terminal'], 'Centro': ['Almacen', 'ZonaNorte', 'Mercado'], 'Terminal': ['Almacen', 'Mercado'], 'Mercado': ['Centro', 'Terminal', 'ZonaNorte'], 'ZonaNorte': ['Centro', 'Mercado']}


In [17]:
def dfs(grafo, inicio, objetivo):
    """
    Busca un camino entre 'inicio' y 'objetivo'
    usando búsqueda en profundidad (DFS).
    """

    pila = [[inicio]]
    visitados = set()

    while pila:
        # Sacar el último camino de la pila
        camino = pila.pop()

        # Obtener el último nodo del camino
        nodo_actual = camino[-1]

        # Comprobar si llegamos al objetivo
        if nodo_actual == objetivo:
            return camino

        # Si todavía no fue visitado
        if nodo_actual not in visitados:
            visitados.add(nodo_actual)

            # Revisar sus vecinos
            for vecino in grafo[nodo_actual]:
                nuevo_camino = camino + [vecino]

                # Agregar el nuevo camino a la pila
                pila.append(nuevo_camino)

    return None

In [18]:
resultado_dfs = dfs(
    mapa,
    "Almacen",
    "ZonaNorte"
)

print(
    "Camino encontrado con DFS:",
    resultado_dfs
)

Camino encontrado con DFS: ['Almacen', 'Terminal', 'Mercado', 'ZonaNorte']


### Comparación entre BFS y DFS

BFS encontró el camino:

`Almacen → Centro → ZonaNorte`

con **2 desplazamientos**.

DFS encontró el camino:

`Almacen → Terminal → Mercado → ZonaNorte`

con **3 desplazamientos**.

BFS utiliza una cola FIFO y explora los nodos por niveles, mientras que DFS utiliza una pila LIFO y profundiza primero en un camino.

En este caso, BFS encontró el camino con menor número de pasos.

## 4.4 Vectores y matrices con NumPy

Un vector representa las características de una sola observación.

En este ejemplo, cada observación contiene dos características:

- Temperatura
- Humedad

Al juntar varias observaciones se obtiene una matriz, donde cada fila representa una observación y cada columna una característica.

### Parte A: Creación de vectores y matriz

In [19]:
import numpy as np
obs1 = np.array([22.5, 60]) # temperatura, humedad
obs2 = np.array([19.0, 45])
obs3 = np.array([25.3, 70])
obs4 = np.array([21.0, 50])
dataset = np.array([obs1, obs2, obs3, obs4])
print(dataset)
print("Forma de la matriz (filas, columnas):", dataset.shape)

[[22.5 60. ]
 [19.  45. ]
 [25.3 70. ]
 [21.  50. ]]
Forma de la matriz (filas, columnas): (4, 2)


### Parte B: Acceso a filas y columnas

Podemos acceder a una observación completa seleccionando una fila, o a una característica específica seleccionando una columna.

In [20]:
print(
    "Primera observación (fila 0):",
    dataset[0]
)

Primera observación (fila 0): [22.5 60. ]


In [21]:
print(
    "Todas las Temperaturas (columna 0):",
    dataset[:, 0]
)

Todas las Temperaturas (columna 0): [22.5 19.  25.3 21. ]


## 4.5 Producto escalar y distancia euclidiana

En esta sección se comparan dos observaciones utilizando dos operaciones matemáticas:

- Producto escalar.
- Distancia euclidiana.

Las observaciones utilizadas son `obs1` y `obs2`.

### Parte A: Producto escalar

El producto escalar combina los valores de dos vectores multiplicando sus componentes correspondientes y sumando los resultados.

In [22]:
producto = np.dot(obs1, obs2)
print("Producto escalar entre obs1 y obs2:", producto)

Producto escalar entre obs1 y obs2: 3127.5


### Parte B: Distancia euclidiana

La distancia euclidiana permite medir qué tan separadas se encuentran dos observaciones.

Mientras menor sea la distancia, más parecidas son las observaciones.

In [23]:
# Formula manual
distancia_manual = np.sqrt(np.sum((obs1 - obs2) ** 2))
print("Distancia manual:", distancia_manual)
# Con la funcion de NumPy
distancia_numpy = np.linalg.norm(obs1 - obs2)
print("Distancia con NumPy:", distancia_numpy)

Distancia manual: 15.402921800749363
Distancia con NumPy: 15.402921800749363


## 5.3 Parte B: Encontrar la observación más cercana

Se desea comparar una nueva observación con las observaciones existentes del dataset.

La nueva observación contiene:

- Temperatura: 20.5 °C
- Humedad: 48 %

Para determinar cuál observación es más parecida, se calcula la distancia euclidiana entre la nueva observación y cada fila del dataset.

La observación con la menor distancia será considerada la más cercana.

In [24]:
nueva_observacion = np.array([20.5, 48])

print("Nueva observación:", nueva_observacion)

Nueva observación: [20.5 48. ]


In [25]:
distancias = []

for i, obs in enumerate(dataset):
    d = np.linalg.norm(nueva_observacion - obs)

    distancias.append(d)

    print(
        f"Distancia a la observación {i}: {d:.4f}"
    )

Distancia a la observación 0: 12.1655
Distancia a la observación 1: 3.3541
Distancia a la observación 2: 22.5175
Distancia a la observación 3: 2.0616


In [26]:
indice_mas_cercano = np.argmin(distancias)

print(
    "La observación más parecida es la número:",
    indice_mas_cercano
)

La observación más parecida es la número: 3


In [27]:
print(
    "Observación más cercana:",
    dataset[indice_mas_cercano]
)

print(
    "Distancia mínima:",
    distancias[indice_mas_cercano]
)

Observación más cercana: [21. 50.]
Distancia mínima: 2.0615528128088303


### Interpretación del resultado

La observación más cercana a `[20.5, 48]` es la observación número 3, correspondiente a `[21.0, 50.0]`.

Su distancia euclidiana es aproximadamente `2.0616`, siendo la menor de todas las distancias calculadas. Por lo tanto, es la observación más similar dentro del dataset.

## 5.4 Reflexión

La distancia euclidiana permite medir qué tan similares son dos observaciones. Cuando la distancia es pequeña, significa que sus características tienen valores parecidos; mientras que una distancia mayor indica que son más diferentes.

Sin embargo, puede no ser adecuada cuando las características tienen escalas muy distintas. Por ejemplo, una variable medida en soles podría tener valores mucho mayores que otra medida en años y dominar el cálculo. En estos casos es recomendable normalizar o estandarizar los datos antes de calcular las distancias.